client/dashboard/utils/domain.ts

```diff
@@ -0,0 +1,54 @@
+import { DotcomFeatures } from '../data/constants';
+import { DomainTypes } from '../data/domains';
+import { UserFlags } from '../data/me';
+import { hasPlanFeature } from './site-features';
+import { userHasFlag } from './user';
+import type { Domain, Site, User } from '../data/types';
+
+export function isDomainRenewable( domain: Domain ) {
+	// Only registered domains can be manually renewed
+	if ( domain.type !== DomainTypes.REGISTERED ) {
+		return false;
+	}
+
+	const shouldNotAllowManualRenew =
+		! domain.subscription_id ||
+		domain.pending_renewal ||
+		domain.pending_registration_at_registry ||
+		domain.pending_registration ||
+		! domain.current_user_can_manage ||
+		( ! domain.is_renewable && ! domain.is_redeemable ) ||
+		domain.aftermarket_auction;
+	return ! shouldNotAllowManualRenew;
+}
+
+export const shouldUpgradeToMakeDomainPrimary = ( {
+	domain,
+	site,
+	user,
+}: {
+	domain: Domain;
+	site?: Site;
```

Comment:
Sort of a nit. This utility function doesn't need to handle the site being undefined.

```suggestion
	site: Site;
```

It's sort of a nit, but also just a point that since we're making good use of React Suspense, it seems like we shouldn't need as many `?`s in our code.

Reply:
Sounds good!